File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/test_descriptor/TestDescriptorResolverRegistry.java
package com.teamscale.test_impacted.test_descriptor;

import com.teamscale.test_impacted.commons.LoggerUtils;
import org.junit.platform.commons.util.ClassLoaderUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.ServiceLoader;
import java.util.logging.Logger;

/**
 * Registry containing the default and custom {@link ITestDescriptorResolver}s discovered by the java
 * {@link ServiceLoader}.
 */
public class TestDescriptorResolverRegistry {

	private static final Logger LOGGER = LoggerUtils.getLogger(TestDescriptorResolverRegistry.class);

	private static final Map<String, ITestDescriptorResolver> TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID = new HashMap<>();

	static {
		// Register default test descriptor resolvers
		registerTestDescriptorResolver(new JUnitJupiterTestDescriptorResolver());
		registerTestDescriptorResolver(new JUnitVintageTestDescriptorResolver());
		registerTestDescriptorResolver(new JUnitPlatformSuiteDescriptorResolver());
		registerTestDescriptorResolver(new CucumberPickleDescriptorResolver());

		// Override existing or register new test descriptor resolvers
		for (ITestDescriptorResolver testDescriptorResolver : ServiceLoader
				.load(ITestDescriptorResolver.class, ClassLoaderUtils.getDefaultClassLoader())) {
			registerTestDescriptorResolver(testDescriptorResolver);
		}
	}

	private static void registerTestDescriptorResolver(ITestDescriptorResolver testDescriptorResolver) {
		TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID.put(testDescriptorResolver.getEngineId(), testDescriptorResolver);
	}

	/** Returns the test descriptor resolver or null if none exists for the test engine. */
	public static ITestDescriptorResolver getTestDescriptorResolver(String testEngineId) {
		if (!TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID.containsKey(testEngineId)) {
			LOGGER.warning(() -> testEngineId + " is not officially supported! You can add support by " +
					"implementing the ITestDescriptorResolver interface and making the implementation " +
					"discoverable via the Java Service Loader mechanism!");
			return TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID.get("junit-jupiter");
		}
		return TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID.get(testEngineId);
	}

}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/test_descriptor/TestDescriptorResolverRegistry.kt
package com.teamscale.test_impacted.test_descriptor

import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import com.teamscale.test_impacted.commons.LoggerUtils.getLogger
import org.junit.platform.commons.util.ClassLoaderUtils
import java.util.*

/**
 * Registry containing the default and custom [ITestDescriptorResolver]s discovered by the java
 * [ServiceLoader].
 */
object TestDescriptorResolverRegistry {
	private val LOGGER = createLogger()

	private val TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID = mutableMapOf<String, ITestDescriptorResolver>()

	init {
		// Register default test descriptor resolvers
		JUnitJupiterTestDescriptorResolver().register()
		JUnitVintageTestDescriptorResolver().register()
		JUnitPlatformSuiteDescriptorResolver().register()
		CucumberPickleDescriptorResolver().register()

		// Override existing or register new test descriptor resolvers
		ServiceLoader
			.load(ITestDescriptorResolver::class.java, ClassLoaderUtils.getDefaultClassLoader())
			.forEach { testDescriptorResolver ->
				testDescriptorResolver.register()
			}
	}

	private fun ITestDescriptorResolver.register() {
		TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID[engineId] = this
	}

	/** Returns the test descriptor resolver or null if none exists for the test engine.  */
	@JvmStatic
	fun getTestDescriptorResolver(testEngineId: String): ITestDescriptorResolver? {
		if (!TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID.containsKey(testEngineId)) {
			LOGGER.warning {
				testEngineId + " is not officially supported! You can add support by " +
						"implementing the ITestDescriptorResolver interface and making the implementation " +
						"discoverable via the Java Service Loader mechanism!"
			}
			return TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID["junit-jupiter"]
		}
		return TEST_DESCRIPTOR_RESOLVER_BY_ENGINE_ID[testEngineId]
	}
}
